File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    private void store(TreeNode root, ArrayList<TreeNode> space) {
        
        if(root == null) return;
        
        space.add(root);
        store(root.left, space);
        store(root.right, space);
    }
    
    public void flatten(TreeNode root) {
        ArrayList<TreeNode> space = new ArrayList<>();
        store(root, space);
        
        int size = space.size() - 1;
        int i = 1;
        while(size --> 0) {
            root.right = space.get(i);
            root.left = null;
            root = root.right;
            i++;
        }
        
    }
}

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if(root == nullptr) return;
        flatten(root->left);
        flatten(root->right);
        
        if(root->left != nullptr) {
            TreeNode* right = root->left;
            
            while(right->right != nullptr)
                right = right->right;
            
            right->right = root->right;
            root->right = root->left;
            root->left = nullptr;
            
        }
    }
};